File: array/two-sum.c
#include <bits/stdc++.h>
#include "vector"

using namespace std;

vector<int> two_sum(vector<int> &nums, int target) {
    map<int, int> table;
    for (int i = 0; i < nums.size(); i++) {
        int diff = target - nums[i];
        if (table.count(nums[i])) {
            return {table[nums[i]], i};
        }
        table[diff] = i;
    }
    return {0, 0};
}

int main() {
    vector<int> nums = {3, 2, 4};

    vector<int> res = two_sum(nums, 6);
    for (auto e: res) {
        cout << e << endl;
    }

}

File: array/two-sum.py
"""
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.
"""

def two_sum(nums, target):
    table = {}
    for i in range(len(nums)):
        diff = target - nums[i]
        if nums[i] in table:
            return [i, table[nums[i]]]
        table[diff] = i


assert two_sum([-1, -2, -3, -4, -5], -8) == [4, 2]
assert two_sum([3, 2, 4], 6) == [2, 1]
assert two_sum([3, 3], 6) == [1, 0]
